upgrade helm sample for 1.0.0 and add script to automate the process (#139)
operator-framework/operator-sdk-samples: helm/README.md

# Memcached Helm Operator

## Overview

This Memcached operator is a simple example of the Operator SDK Helm-based operator. It is based on the [`stable/memcached` chart][stable/memcached] .

## Prerequisites

- [docker][docker_tool] version 17.03+
- [kubectl][kubectl_tool] v1.14.1+
- [operator SDK][operator_install]
- Access to a Kubernetes v1.16.0+ cluster.

## Getting Started

### Clone the repository

Checkout this Memcached operator repository

> ```
> $ git clone https://github.com/operator-framework/operator-sdk-samples.git
> $ cd operator-sdk-samples/helm/memcached-operator

Reviewer: ````suggestion
```sh
git clone https://github.com/operator-framework/operator-sdk-samples.git
cd operator-sdk-samples/helm/memcached-operator
````

```

### Build the operator image

Build the Memcached operator image and push it to a public registry, such as quay.io:

> ```

Reviewer: ````suggestion
```sh
````

export IMAGE=quay.io/example-inc/memcached-operator:v0.0.1
make docker-build docker-push IMG=$IMAGE
```

**NOTE** To allow the cluster pull the image the repository needs to be set as public or you must configure an image pull secret.

### Run the operator

Deploy the project to the cluster. Set `IMG` with `make deploy` to use the image you just pushed:

> ```

Reviewer: ````suggestion
```sh
````

make deploy IMG=$IMAGE
```

### Create a sample custom resource

Create a sample CR:

```sh
kubectl apply -f config/samples/cache_v1alpha1_memcached.yaml -n memcached-operator-system
```

Run the following command to verify that the installation was successful:

> ```sh
> kubectl get all -n memcached-operator-system

Reviewer: ````suggestion
```console
$ kubectl get all -n memcached-operator-system
````

> kubectl get all -n memcached-operator-system

Reviewer: ```suggestion
$ kubectl get all -n memcached-operator-system
```

NAME                                                    READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/memcached-operator-controller-manager   1/1     1            1           2m16s

NAME                                                              DESIRED   CURRENT   READY   AGE
replicaset.apps/memcached-operator-controller-manager-d54b5fb78   1         1         1       2m16s

NAME                                READY   AGE
statefulset.apps/memcached-sample   3/3     96s
```

### Clean up

Delete the CR to uninstall the release:
`
```sh
kubectl delete -f config/samples/cache_v1alpha1_memcached.yaml -n memcached-operator-system
````

Use `make uninstall` and `make undeploy` to uninstall the operator and its CRDs:

```sh
make uninstall
make undeploy
```

### Troubleshooting

Run the following command to check the operator logs.

```sh
kubectl logs deployment.apps/memcached-operator-controller-manager -n memcached-operator-system -c manager
```

### Extras

This project was created by using the [gen-helm-memcached.sh][gen-helm-memcached.sh] script , which means that it is using the official [stable/memcached][stable/memcached] helm chart.

> Note that you must have Helm installed locally and add the stable helm charts repo to it to use the `stable/memcached` Helm charts. See the [Helm Quickstart guide][helm-quick] for installation instructions.

Author: ```suggestion
Note that you must have Helm installed locally and add the stable helm charts repo to it to use the `stable/memcached` Helm charts. See the [Helm Quickstart guide][helm-quick] for installation instructions. Also, you can start with Helm-based Operators with SDK by checking its [quickstart][helm_guide].
```

[kubectl_tool]: https://kubernetes.io/docs/tasks/tools/install-kubectl/
[docker_tool]: https://docs.docker.com/install/
[operator_install]: https://sdk.operatorframework.io/docs/install-operator-sdk/
[helm_guide]: https://sdk.operatorframework.io/docs/helm/quickstart/
[stable/memcached]: https://github.com/helm/charts/tree/master/stable/memcached
[helm-quick]: https://helm.sh/docs/intro/quickstart/

> [gen-helm-memcached.sh]: .generate/gen-helm-sample.sh

Reviewer 2: ```suggestion
[gen-helm-memcached.sh]: .generate/gen-helm-memcached.sh
```

Author: ?

# Memcached Helm Operator

## Overview

This Memcached operator is a simple example of the Operator SDK Helm-based operator. It is based on the [`stable/memcached` chart][stable/memcached] .

## Prerequisites

- [docker][docker_tool] version 17.03+
- [kubectl][kubectl_tool] v1.14.1+
- [operator SDK][operator_install]
- Access to a Kubernetes v1.16.0+ cluster.

## Getting Started

### Clone the repository

Checkout this Memcached operator repository

```

> $ mkdir operator-framework
> $ cd operator-framework

Reviewer 2: ```suggestion
```

$ git clone https://github.com/operator-framework/operator-sdk-samples.git
$ cd operator-sdk-samples/helm/memcached-operator
```

### Build the operator image

Build the Memcached operator image and push it to a public registry, such as quay.io:

```

> export IMAGE=quay.io/example-inc/memcached-operator:v0.0.1
> make docker-build docker-push IMG=$IMAGE

Reviewer 2: ```suggestion
export IMG=quay.io/example-inc/memcached-operator:v0.0.1
make docker-build docker-push
```

```

**NOTE** To allow the cluster pull the image the repository needs to be set as public or you must configure an image pull secret.

### Run the operator

> Install the CRD and deploy the project to the cluster. Set IMG with make deploy to use the image you just pushed:

Reviewer 2: ```suggestion
Install the CRD and deploy the project to the cluster. Set `IMG` with make deploy to use the image you just pushed:
```

```

> make install

Reviewer 2: ```suggestion
```
`make deploy` installs the CRDs, `make install` is only necessary when running locally.

make deploy IMG=$IMAGE
```

### Create a sample custom resource

Create a sample CR:

```sh
kubectl apply -f config/samples/cache_v1alpha1_memcached.yaml -n memcached-operator-system
```

Run the following command to verify that the installation was successful:

```sh

> $ kubectl get all -n memcached-operator-system

Reviewer 2: ```suggestion
kubectl get all -n memcached-operator-system
```

Reviewer: prefixing commands with `$` is pretty common in SDK docs since it differentiates command output from the commands themselves. If we're using `sh` blocks _without output_, removing `$` makes sense. Otherwise we should be using `$ cmd` and `console` code blocks.

Author: Make sense 👍

service/memcached-operator-controller-manager-metrics-service   ClusterIP   10.107.115.48   <none>        8443/TCP    2m16s
service/memcached-sample                                        ClusterIP   None            <none>        11211/TCP   96s

NAME                                                    READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/memcached-operator-controller-manager   1/1     1            1           2m16s

NAME                                                              DESIRED   CURRENT   READY   AGE
replicaset.apps/memcached-operator-controller-manager-d54b5fb78   1         1         1       2m16s

NAME                                READY   AGE
statefulset.apps/memcached-sample   3/3     96s
```

### Clean up

Delete the CR to uninstall the release:
`
```sh
kubectl delete -f config/samples/cache_v1alpha1_memcached.yaml -n memcached-operator-system
````

Use `make uninstall` and `make undeploy` to uninstall the operator and its CRDs:

```sh
make uninstall
make undeploy
```

### Troubleshooting

Run the following command to check the operator logs.

> ```shell

Reviewer 2: ```suggestion
```sh
```
supernit for consistency

kubectl logs deployment.apps/memcached-operator-controller-manager -n memcached-operator-system -c manager
```

### Extras

> This project was created by using the [gen-helm-sample.sh][gen-helm-sample.sh] script , which means that it is using the official [stable/memcached][stable/memcached] helm chart.

Reviewer 2: ```suggestion
This project was created by using the [gen-helm-memcached.sh][gen-helm-memcached.sh] script , which means that it is using the official [stable/memcached][stable/memcached] helm chart.
```

service/memcached-sample                                        ClusterIP   None            <none>        11211/TCP   96s

NAME                                                    READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/memcached-operator-controller-manager   1/1     1            1           2m16s

NAME                                                              DESIRED   CURRENT   READY   AGE
replicaset.apps/memcached-operator-controller-manager-d54b5fb78   1         1         1       2m16s

NAME                                READY   AGE
statefulset.apps/memcached-sample   3/3     96s
```

### Clean up

Delete the CR to uninstall the release:
`
```sh
kubectl delete -f config/samples/cache_v1alpha1_memcached.yaml -n memcached-operator-system
````

Use `make uninstall` and `make undeploy` to uninstall the operator and its CRDs:

```sh
make uninstall
make undeploy
```

### Troubleshooting

Run the following command to check the operator logs.

```shell
kubectl logs deployment.apps/memcached-operator-controller-manager -n memcached-operator-system -c manager
```

> **NOTE** For further information about how to develop Helm operators with Operator SDK, read the [Helm User Guide for Operator SDK][helm_guide]

Reviewer 2: I think this link should be the tutorial rather than the quickstart.

Author: The quick will lead the user to the tutorial so I think it is fine we keep te quick. Mainly because it might be the first user contact with. But, I think we should change the info here as well. See^

> **NOTE** For further information about how to develop Helm operators with Operator SDK, read the [Helm User Guide for Operator SDK][helm_guide]

Author: ```suggestion
**NOTE** For further information about how to develop Helm-based Operators with Operator SDK, getting started by checking the [Helm QuickStart Operator SDK][helm_guide].
```

> **NOTE** For further information about how to develop Helm-based Operators with Operator SDK, getting started by checking the [Helm QuickStart Operator SDK][helm_guide].

Author: ```suggestion
**NOTE** For further information about how to develop Helm-based Operators with Operator SDK, getting started by checking the [Helm Quickstart Operator SDK][helm_guide].
```
